### PR TITLE
Training Workload for Gemma 3 1B  - LoRA Finetuning

### LINKED ISSUE #267: ☂️ Training workload for google/gemma-3-1b-it with LoRA
Train google/gemma-3-1b-it using technique LoRA as a workaround due to unsupported Ops for full-model training.

### PR DESCRIPTION
**Pull request by pglusacTT**
### Ticket
Closes #267

### Model training setting

- **Model:** `Gemma 3 1B instruction tuned`
- **Strategy:** SFT with LoRA layers on a single chip (N150)
- **Framework:** PyTorch
- **Task:** Sentiment prediction using `SST2`

### What's changed
Implement training workload for Gemma 3 1B

### Experiment results
image

Examples:

Step | Input Text | Target Text | Pred Text
-- | -- | -- | --
1 | Review: in exactly 89 minutes , most of which passed as slowly as if i 'd been sitting naked on an igloo , formula 51 sank from quirky to jerky to utter turkey . | {"label": "negative"} | ": "narr",
150 | Review: or doing last year 's taxes with your ex-wife . | {"label": "negative"} | {"status": "negative"}
600 | Review: a sometimes tedious film . | {"label": "negative"} | {"label": "negative"}

### Checklist
- [x] Main training workflow
- [x] W&B logging support
- [x] CPU testing
- [x] GPU testing
- [x] TT device testing

### CODE REVIEW

File: blacksmith/experiments/torch/gemma/test_gemma_finetuning.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+# Logging settings
+log_level: "INFO"
+use_wandb: True
+wandb_project: "gemma-finetuning"
+wandb_run_name: "tt-gemma-test"
+wandb_tags: ["test"]
+wandb_watch_mode: "all"
+wandb_log_freq: 1000
+model_to_wandb: False
+log_steps_freq: 10
```
Comment (vladimirjovanovicTT) [inline blacksmith/experiments/torch/gemma/test_gemma_finetuning.yaml]: should this be just steps_freq?

File: blacksmith/experiments/torch/gemma/test_gemma_finetuning.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+                # Backward pass
+                loss.backward()
+
+                # Update parameters
+                if config.use_tt:
+                    xm.optimizer_step(optimizer)
+                    torch_xla.sync(wait=True)
+                else:
+                    optimizer.step()
```
[the remark this answers is not in the document]
Comment (pglusacTT) [reply blacksmith/experiments/torch/gemma/test_gemma_finetuning.py]: It's running successfully on device.

File: blacksmith/experiments/torch/gemma/test_gemma_finetuning.yaml [hunk 3]
```diff
@@ -0,0 +1,57 @@
+# Dataset settings
+dataset_id: "stanfordnlp/sst2"
+
+# Model settings
+model_name: "google/gemma-3-1b-it"
+max_length: 32
```
[the remark this answers is not in the document]
Comment (pglusacTT) [reply blacksmith/experiments/torch/gemma/test_gemma_finetuning.yaml]: I realized that this dataset doesn't really require 64, so I set 32 to reduce memory usage.

File: blacksmith/experiments/torch/gemma/test_gemma_finetuning.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+dtype: "torch.bfloat16"
+
+# Training hyperparameters
+learning_rate: 6e-5
+batch_size: 8
```
[the remark this answers is not in the document]
Comment (pglusacTT) [reply blacksmith/experiments/torch/gemma/test_gemma_finetuning.yaml]: This is something I've left when debugging, I'll see if I can maybe set it to a higher number.
